In [ ]:
import pandas as pd
import numpy as np
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine

# Global Commerce Intelligence Platform (GCIP) V1

## Exploratory Data Analysis (EDA)

**Objective:**
Connect to PostgreSQL, retrieve data using SQL, perform exploratory data analysis, and identify business insights to support dashboard development.

In [ ]:
# Display all columns
pd.set_option('display.max_columns', None)

# Display wider DataFrames
pd.set_option('display.width', 1000)

print("Libraries Imported Successfully."
      )

Libraries Imported Successfully.


In [ ]:
load_dotenv()

#loaded enviroment variables 

True

In [ ]:
# connected to postgresql 

engine = create_engine( f"postgresql+psycopg2://{os.getenv('DB_USER')}:{os.getenv('DB_PASSWORD')}@{os.getenv('DB_HOST')}:{os.getenv('DB_PORT')}/{os.getenv('DB_NAME')}")
print("Database Connection Established Successfully!")

Database Connection Established Successfully!


In [ ]:
#testing connection
query = """
SELECT current_database(),
       current_user;
"""

connection_test = pd.read_sql(query, engine)

connection_test

,current_database,current_user
0,gcip_db,postgres


In [ ]:
customers = pd.read_sql("SELECT * FROM customers;", engine)
orders = pd.read_sql("SELECT * FROM orders;", engine)
order_items = pd.read_sql("SELECT * FROM order_items;", engine)
products = pd.read_sql("SELECT * FROM products;", engine)
sellers = pd.read_sql("SELECT * FROM sellers;", engine)
order_payments = pd.read_sql("SELECT * FROM order_payments;", engine)
order_reviews = pd.read_sql("SELECT * FROM order_reviews;", engine)
geolocation = pd.read_sql("SELECT * FROM geolocation;", engine)
product_category_name_translation = pd.read_sql(
    "SELECT * FROM product_category_name_translation;", engine
)

In [ ]:
tables = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "products": products,
    "sellers": sellers,
    "order_payments": order_payments,
    "order_reviews": order_reviews,
    "geolocation": geolocation,
    "product_category_name_translation": product_category_name_translation
}

In [ ]:
customers.head()
customers.shape #(99441, 5)
customers.info()
customers.columns

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state'], dtype='str')

In [ ]:
customers.describe(include="all")
customers.duplicated().sum()
customers.isnull().sum()
customers.nunique()



customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

### Observations

- Dataset contains 99441 customer records(rows).
- No missing values were found.
- No duplicate rows were detected.
- Customer IDs are stored correctly as object datatype.
- Data appears clean and ready for analysis.

In [ ]:
#orders table 
orders.shape  #(99441,8)
orders.head()
# orders.info()
orders.describe(include="all")
orders.duplicated().sum() #0
orders.isnull().sum() #null values are present in these columns mentioned below 
# order_approved_at                 160
# order_delivered_carrier_date     1783    
# order_delivered_customer_date    2965
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

## observations
* orders table have (99441,8)(rows , column)
* no duplicate value exists
* missing values found in these columns mentioned below 
order_approved_at                 160
 order_delivered_carrier_date     1783    
 order_delivered_customer_date    2965

In [ ]:
#order_items

In [ ]:
order_items.shape #(112650 ,7)
order_items.head()
order_items.info()
order_items.describe(include="all")
order_items.duplicated().sum() #0
order_items.isnull().sum()#0
order_items.nunique()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  str           
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  str           
 3   seller_id            112650 non-null  str           
 4   shipping_limit_date  112650 non-null  datetime64[us]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(3)
memory usage: 6.0 MB


order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [ ]:
#products 
products.shape #(32951 , 9 )
products.head()
products.info()
products.describe(include="all")
products.duplicated().sum() #0
products.isnull().sum() 
# products.nunique()

<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_length         32341 non-null  float64
 3   product_description_length  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB


product_id                      0
product_category_name         610
product_name_length           610
product_description_length    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

## observatiosn



* null vlaues exists here 
8product_id                      0
product_category_name         610
product_name_length           610
product_description_length    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2

In [ ]:
products[products.isnull().any(axis=1)]

,product_id,product_category_name,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.0,17.0,14.0,12.0
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.0,16.0,7.0,20.0
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.0,20.0,20.0,20.0
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,18500.0,41.0,30.0,41.0
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.0,35.0,7.0,12.0
...,...,...,...,...,...,...,...,...,...
32515,b0a0c5dd78e644373b199380612c350a,NaN,NaN,NaN,NaN,1800.0,30.0,20.0,70.0
32589,10dbe0fbaa2c505123c17fdc34a63c56,NaN,NaN,NaN,NaN,800.0,30.0,10.0,23.0
32616,bd2ada37b58ae94cc838b9c0569fecd8,NaN,NaN,NaN,NaN,200.0,21.0,8.0,16.0
32772,fa51e914046aab32764c41356b9d4ea4,NaN,NaN,NaN,NaN,1300.0,45.0,16.0,45.0


In [ ]:
products[
    products[
        ['product_category_name',
         'product_name_length',
         'product_description_length',
         'product_photos_qty']
    ].isnull().all(axis=1)
]

,product_id,product_category_name,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.0,17.0,14.0,12.0
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.0,16.0,7.0,20.0
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.0,20.0,20.0,20.0
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,18500.0,41.0,30.0,41.0
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.0,35.0,7.0,12.0
...,...,...,...,...,...,...,...,...,...
32515,b0a0c5dd78e644373b199380612c350a,NaN,NaN,NaN,NaN,1800.0,30.0,20.0,70.0
32589,10dbe0fbaa2c505123c17fdc34a63c56,NaN,NaN,NaN,NaN,800.0,30.0,10.0,23.0
32616,bd2ada37b58ae94cc838b9c0569fecd8,NaN,NaN,NaN,NaN,200.0,21.0,8.0,16.0
32772,fa51e914046aab32764c41356b9d4ea4,NaN,NaN,NaN,NaN,1300.0,45.0,16.0,45.0


In [ ]:
products[
    products[
        ['product_weight_g',
         'product_length_cm',
         'product_height_cm',
         'product_width_cm']
    ].isnull().all(axis=1)
]

,product_id,product_category_name,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
8582,09ff539a621711667c43eba6a3bd8466,bebes,60.0,865.0,3.0,NaN,NaN,NaN,NaN
18852,5eb564652db742ff8f28759cd8d2652a,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
missing_products = products[products.isnull().any(axis=1)]

order_items[
    order_items["product_id"].isin(missing_products["product_id"])
]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
123,0046e1d57f4c07c8c92ab26be8c3dfc0,1,ff6caf9340512b8bf6d2a2a6df032cfa,38e6dada03429a47197d5d584d793b41,2017-10-02 15:49:17,7.79,7.78
125,00482f2670787292280e0a8153d82467,1,a9c404971d1a5b1cbc2e4070e02731fd,702835e4b785b67a084280efca355756,2017-02-17 16:18:07,7.60,10.96
132,004f5d8f238e8908e6864b874eda3391,1,5a848e4ab52fd5445cdc07aab1c40e48,c826c40d7b19f62a09e2d7c5e7295ee2,2018-03-06 09:29:25,122.99,15.61
142,0057199db02d1a5ef41bacbf41f8f63b,1,41eee23c25f7a574dfaf8d5c151dbb12,e5a3438891c0bfdb9394643f95273d8e,2018-01-25 09:07:51,20.30,16.79
171,006cb7cafc99b29548d4f412c7f9f493,1,e10758160da97891c2fdcbc35f0f031d,323ce52b5b81df2cd804b017b7f09aa7,2018-02-22 13:35:28,56.00,14.14
...,...,...,...,...,...,...,...
112306,ff24fec69b7f3d30f9dc1ab3aee7c179,1,5a848e4ab52fd5445cdc07aab1c40e48,c826c40d7b19f62a09e2d7c5e7295ee2,2018-02-01 02:40:12,122.99,15.61
112333,ff3024474be86400847879103757d1fd,1,f9b1795281ce51b1cf39ef6d101ae8ab,3771c85bac139d2344864ede5d9341e3,2017-11-21 03:55:39,39.90,9.94
112350,ff3a45ee744a7c1f8096d2e72c1a23e4,1,b61d1388a17e3f547d2bc218df02335b,07017df32dc5f2f1d2801e579548d620,2017-05-10 10:15:19,139.00,21.42
112438,ff7b636282b98e0aa524264b295ed928,1,431df35e52c10451171d8037482eeb43,6cd68b3ed6d59aaa9fece558ad360c0a,2018-02-22 15:35:35,49.90,15.11


## observation
 * 610 products have missing product information.
* 2 products have missing physical dimensions.
* These products appear in 1,604 order item records, indicating they were sold.
* Therefore, the missing values should be documented and handled carefully during * analysis rather than removing the records.

In [ ]:
order_payments.shape  #103886 ,5
order_payments.head()
order_payments.info()
order_payments.describe(include="all")
order_payments.duplicated().sum() #0
order_payments.isnull().sum() #0
order_payments.nunique()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

In [ ]:
order_reviews.shape # 99224,7
order_reviews.head()
order_reviews.info()
order_reviews.describe(include="all")
order_reviews.duplicated().sum() #0
order_reviews.isnull().sum()
order_reviews.nunique()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  str           
 1   order_id                 99224 non-null  str           
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  str           
 4   review_comment_message   40977 non-null  str           
 5   review_creation_date     99224 non-null  datetime64[us]
 6   review_answer_timestamp  99224 non-null  datetime64[us]
dtypes: datetime64[us](2), int64(1), str(4)
memory usage: 5.3 MB


review_id                  98410
order_id                   98673
review_score                   5
review_comment_title        4515
review_comment_message     36155
review_creation_date         636
review_answer_timestamp    88867
dtype: int64


* below are the columns having null values
* eview_comment_title       87656
* review_comment_message     58247

In [ ]:
sellers.shape #3095 , 4
sellers.head()
sellers.info()
sellers.describe(include="all")
sellers.duplicated().sum()#0
sellers.isnull().sum()  #0
sellers.nunique()

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB


seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

In [ ]:
geolocation.shape   #1000163 , 5
geolocation.head()
geolocation.info()
geolocation.describe(include="all")
geolocation.duplicated().sum() # 262010 
# noraml duplicate values as state names , location can repeat 
geolocation.isnull().sum() #0
geolocation.nunique()

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 38.2 MB


geolocation_zip_code_prefix     19015
geolocation_lat                658242
geolocation_lng                674800
geolocation_city                 8011
geolocation_state                  27
dtype: int64

In [ ]:
product_category_name_translation.shape #71,2
product_category_name_translation.head()
product_category_name_translation.info()
product_category_name_translation.describe(include="all")
product_category_name_translation.duplicated().sum() #0
product_category_name_translation.isnull().sum() #0
product_category_name_translation.nunique() 

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB


product_category_name            71
product_category_name_english    71
dtype: int64

### Data Cleaning Summary

The dataset was inspected for missing values, duplicates, inconsistent data types, and other data quality issues. Missing values were investigated individually to determine whether they represented data quality problems or legitimate missing information.

No records were removed or imputed because the missing values either represented expected business scenarios (such as missing review comments or delivery dates for undelivered orders) or belonged to products that were associated with valid sales transactions. Preserving these records ensures that business analyses remain accurate and unbiased.